compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java

```diff
@@ -610,6 +618,61 @@ public void visit(BLangInvocation invocationExpr) {
         }
     }
 
+    private boolean isFieldsInitializedForSelfKeyWordExpression(BLangInvocation invocationExpr) {
+
+        if (invocationExpr.expr == null || !isSelfKeyWordExpr(invocationExpr.expr)) {
+            return true;
+        }
+        StringBuilder uninitializedFields =
+                getUninitializedFieldsForSelfKeyword((BObjectType) ((BLangSimpleVarRef)
+                        invocationExpr.expr).symbol.type);
+        if (uninitializedFields.length() != 0) {
+            this.dlog.error(invocationExpr.pos, DiagnosticCode.CONTAINS_UNINITIALIZED_FIELDS,
+                    uninitializedFields.toString());
+            return false;
+        }
+        return true;
+    }
+
+    private boolean isFieldsInitializedForSelfKeyWordMethodInvocation(BLangInvocation invocationExpr) {
+
+        for (BLangExpression expr : invocationExpr.requiredArgs) {
+            if (isSelfKeyWordExpr(expr)) {
+                StringBuilder uninitializedFields =
+                        getUninitializedFieldsForSelfKeyword((BObjectType) ((BLangSimpleVarRef) expr).symbol.type);
+                if (uninitializedFields.length() != 0) {
+                    this.dlog.error(invocationExpr.pos, DiagnosticCode.CONTAINS_UNINITIALIZED_FIELDS,
+                            uninitializedFields.toString());
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    private boolean isSelfKeyWordExpr(BLangExpression expr) {
+
+        return expr.getKind() == NodeKind.SIMPLE_VARIABLE_REF &&
+                Names.SELF.value.equals(((BLangSimpleVarRef) expr).getVariableName().getValue());
+    }
+
+    private StringBuilder getUninitializedFieldsForSelfKeyword(BObjectType objType) {
+
+        boolean isFirstUninitializedField = true;
+        StringBuilder uninitializedFields = new StringBuilder();
+        for (BField field : objType.fields) {
+            if (this.uninitializedVars.containsKey(field.symbol)) {
```

Comment:
Aren't we checking if the fields are initialized? The following should NOT fail.

```ballerina
import ballerina/io;

type Foo object {
    string s;

    public function __init(string s) {
        self.s = s;
        self.printS(); // no problem since self.s is initialized
    }

    function printS() {
        io:println(self.s);
    }
};
```

Reply:
Since `this.uninitializedVars` only contains uninitialized variables do we need to check that?

Reply:
Since `this.uninitializedVars` only contains the uninitilized variables Do we need to check that?

Reply:
Ah yeah, sorry misread this. This should be enough.

Shall we add a test though, for the same?